File: Shinpei2-cs50-2019-x-credit-20190709T012653Z/credit.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <stdlib.h>

int checkDigit(long number);
bool Luhn(long number);

int main(void)
{       
    //カード番号を入力(Luhnチェックが真の場合のみ、カード会社を探索)
    long number = get_long("Number: ");
    if (Luhn(number))
    {
        //カード番号の桁数を取得
        int len = checkDigit(number);
        //printf("len:%i\n",len);
        

        //1桁目、2桁目をdigit[0],digit[1]に]格納
        int digit[2];
        long value = number;
        for (int i = 1; i < len - 1; i++)
        {
            value /= 10;
        }
        digit[1] = value % 10;
        value /= 10;
        digit[0] = value % 10;
    
        //printf("1桁目：%d\n",digit[0]);
        //printf("2桁目：%d\n",digit[1]);
    
        //カード番号の1桁目をチェック
        switch (digit[0])
        {
            //AMEXかどうかを調べる。
            case 3:
                if (len == 15 && (digit[1] == 4 || digit[1] == 7))
                {
                    printf("AMEX\n");
                    break;
                }
                printf("INVALID\n");
                break;
              
            case 4:
                if (len == 13 || len == 16)
                {
                    printf("VISA\n");
                    break;
                }
                printf("INVALID\n");
                break;
            case 5:
                if (len == 16 && (1 <= digit[1] && digit[1] <= 5))
                {
                    printf("MASTERCARD\n");
                    break;
                }
                printf("INVALID\n");
                break;
            default:
                printf("INVALID\n");
                break;
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

//カード番号の桁数を調べる関数chckDigit()
int checkDigit(long number)
{
    int digit = 0;
    while (number != 0)
    {
        number = number / 10;
        digit += 1;
    }
    return digit;
}

//クレジットカードの有効/無効を調べる関数Luhn
bool Luhn(long number)
{
    int len = checkDigit(number);
    int odd = 0;
    int even = 0;
    int checkSum = 0;

    //桁数が偶数の場合
    if (len % 2 == 0)
    {
        for (int i = 0; i <= len / 2; i++)
        {
            odd = number % 10;
            number /= 10;
            even = (number % 10) * 2;
            //偶数番目の桁が10以上の場合に変換
            if (even >= 10)
            {
                even = (even % 10) + ((even / 10) % 10);
            }
            checkSum += odd + even;
            number /= 10;
        }
    }
    
    //桁数が奇数の場合
    else
    {
        //最高桁以外のチェックサム合計を計算
        for (int i = 0; i <= len / 2; i++)
        {
            odd = number % 10;
            number /= 10;
            even = (number % 10) * 2;
            //偶数番目の桁が10以上の場合に変換
            if (even >= 10)
            {
                even = (even % 10) + ((even / 10) % 10);
            }
            checkSum += odd + even;
            number /= 10;
        }
        
        //最高位をチェックサムに加算
        checkSum += number % 10;
    }
    
    //printf("checkSum: %i\n",checkSum);
    
    //checkSumが、10で割り切れる場合：true / 割り切れない場合：false
    if (checkSum % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

File: Shinpei2-cs50-2019-x-sentimental-credit-20190727T012507Z/credit.py
from cs50 import get_string


# numberを取得する関数
def get_Number():
    while True:
        number = get_string("Number: ")
        if number.isnumeric():
            if int(number) > 0:
                return number


# checkSumを調べる関数
def checkSum(number, digit):
    cSum = 0
    check = 1
    # チェックサムの計算
    for i in range(digit):
        checkDigit = digit - i - 1
        num = 0
        # 後ろから奇数番目
        if check % 2 == 1:
            num = int(number[checkDigit])
        # 後ろから偶数番目
        else:
            num = int(number[checkDigit]) * 2
        check += 1

        # numが10以上の場合
        if num >= 10:
            cSum += num % 10
            num = num // 10
            cSum += num
        # 10未満の場合
        else:
            cSum += num

    # print(cSum)

    # チェックサム判定
    if cSum % 10 == 0:
        return True
    else:
        return False


def main():
    number = get_Number()

    # numberListとdigitの用意
    numberList = []
    digit = len(number)

    ### チェックサムの判定 ###
    # 桁数が13or15or16の場合
    if digit == 13 or digit == 15 or digit == 16:
        # numberの各桁の文字をnumberListへ格納
        for i in range(2):
            numberList.append(number[i])
        # print(numberList)

        ### クレジットカード会社判定 ###
        # AMEX
        if numberList[0] == '3' and numberList[1] == '4' or numberList[1] == '7' and digit == 15:
            if checkSum(number, digit):
                print("AMEX")
            else:
                print("INVALID")

        # MASTERCARD
        elif numberList[0] == '5' and numberList[1] == '1' or numberList[1] == '2' or numberList[1] == '3' or numberList[1] == '4' or numberList[1] == '5' and digit == 16:
            if checkSum(number, digit):
                print("MASTERCARD")
            else:
                print("INVALID")

        # VISA
        elif numberList[0] == '4' and digit == 13 or digit == 16:
            if checkSum(number, digit):
                print("VISA")
            else:
                print("INVALID")

        # それ以外は無効
        else:
            print("INVALID")
    # それ以外の桁数は無効
    else:
        print("INVALID")


if __name__ == "__main__":
    main()